### PR TITLE
Added ${OnHasProperties} -  for easier custom output

### PR DESCRIPTION
**Pull request by snakefoot**
Can do this (Inspired by [OnException](https://github.com/nlog/NLog/wiki/OnException-Layout-Renderer)):

> `${onhasproperties:, Properties\: ${all-event-properties}}`

Instead of this (performance hit from double rendering):

> `${when:when=length('${all-event-properties}')>0:inner=, Properties\: ${all-event-properties}}`

I guess when #3830 has been merged, then one could also create a static condition-method called `HasProperties(LogEventInfo)`.

Note merge to master

### CODE REVIEW

**Comment by 304NotModified** [comment]
thanks!

I think "on" are event things.

What about "when-has-properties" or "has-properties"

**Comment by snakefoot** [comment]
If going the when-lane, then I would probably just stick with `hasProperties` when you have resolved #3830

**Comment by 304NotModified** [comment]
> If going the when-lane, then I would probably just stick with `hasProperties`-condition-method, **when** you have resolved #3830

I'm missing something. I don't see the relationship to #3830 (also not in https://github.com/NLog/NLog/pull/3833#issue-385206099)). Please elaborate, thanks!

**Comment by snakefoot** [comment]
Then you can do this without allocation:

`${when:when=HasProperties:inner=Properties:}`

Or this:

`${when:when=HasException:inner=Properties:}`

Using condition-methods with single parameter.

File: src/NLog/LayoutRenderers/Wrappers/OnHasPropertiesLayoutRendererWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.LayoutRenderers.Wrappers
+{
+    using System.Text;
+    using NLog.Config;
+
+    /// <summary>
+    /// Outputs alternative layout when the inner layout produces empty result.
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/Wrappers/OnHasPropertiesLayoutRendererWrapper.cs]: Please add the example to the code, thanks!

File: src/NLog/LayoutRenderers/AllEventPropertiesLayoutRenderer.cs [hunk 2]
```diff
@@ -223,7 +223,7 @@ private bool CheckForExclude(LogEventInfo logEvent)
                 checkForExclude = !(logEvent.CallSiteInformation == null && Exclude.Count == CallerInformationAttributeNames.Length && Exclude.Contains(CallerInformationAttributeNames[0]));
             }
 #endif
-            return checkForExclude && logEvent.HasProperties;
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/AllEventPropertiesLayoutRenderer.cs]: double check, this check isn't needed? (probably check in the calling method?)

**Comment by 304NotModified** [comment]
> Then you can do this without allocation:
> 
> `${when:when=HasProperties:inner=Properties:}`
> 
> Or this:
> 
> `${when:when=HasException:inner=Exception:}`
> 
> Using condition-methods with single LogEventInfo-parameter.

Thanks, clear now! (the confusion was somewhere else, but resolved)

File: src/NLog/LayoutRenderers/AllEventPropertiesLayoutRenderer.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/AllEventPropertiesLayoutRenderer.cs]: No it has already been checked before calling this method.

**Comment by snakefoot** [comment]
Think I was too eager with plans for optimizing when-condition-methods. The method-factory will require some refactoring for this to occur. Think I will stick with:

> `${onhasproperties:, Properties\: ${all-event-properties}}`

File: src/NLog/LayoutRenderers/Wrappers/OnHasPropertiesLayoutRendererWrapper.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/Wrappers/OnHasPropertiesLayoutRendererWrapper.cs]: Added example code

Comment (304NotModified) [reply src/NLog/LayoutRenderers/Wrappers/OnHasPropertiesLayoutRendererWrapper.cs]: Thanks!

**Comment by snakefoot** [comment]
Created wiki-page: https://github.com/NLog/NLog/wiki/OnHasProperties-Layout-Renderer

**Comment by snakefoot** [comment]
Create Doc-PR: https://github.com/NLog/NLog.github.io/pull/141

**Pull request merged**